### ISSUE TITLE
Do not require that key ID is provided with the client's public key

### ISSUE THREAD

**Issue by mitar** [ory/fosite#452]
**Is your feature request related to a problem? Please describe.**

It seems clients are bad at managing their keys and making sure key IDs are unique for keys they provide. Also sometimes when they generate a key it simply does not contain the key ID. So they have to manually add it and stuff.

But I think key ID in client's public key should not even be required. RFC says that [`kid` header parameter is optional](https://tools.ietf.org/html/rfc7515#section-4.1.4), [key identification section](https://tools.ietf.org/html/rfc7515#section-6) also talks that any way of figuring out the key can be used, and [appendix D](https://tools.ietf.org/html/rfc7515#appendix-D) explicitly mentions simply trying all possibly viable keys.

**Describe the solution you'd like**

I think `kid` should be optional. It is just an optimization. We could go over all keys client has and try them in order. If `kid` is provided, we can make that faster. But going over all keys should be enough.

Or at least, if there is only one key in the list of client's keys, then the `kid` should be optional. We should just return that one key.

**Describe alternatives you've considered**

There are of course so many other fields we could be selecting keys on, like `x5t`. So making `kid` required is really a very opinionated decision. It could even be `x5t` by which you select the key. So if you want to keep that we do not traverse all keys client has configured (which I doubt is expensive because those lists tend to be short), we could ideally at least use many more mechanisms to match the one key (`kid`, having only one key in the list, `x5t`, and so on). I think this is a bit of an overkill to support all those and I would leave to others to implement that if somebody will want. But making `kid` optional is something I would implement if you ACK it.

There is a bit of an issue that that the [library currently used](https://github.com/dgrijalva/jwt-go) requires only one key to be provided for validation. I opened https://github.com/dgrijalva/jwt-go/issues/416 to address this. Because of this I think that for now the most reasonable way is to allow `kid` to not be provided if there is only one key listed for the client. If so, just return that key if `kid` is not provided.

**Comment by aeneasr** [comment]
Checking against possibly hundreds of RSA keypairs every couple of seconds (assuming medium scale) is a massive CPU drain. We had this problem in ORY Oathkeeper and had to solve it with caching even though we knew the `kid`.

**Comment by mitar** [comment]
So you are saying one particular client would have possibly hundreds of RSA keypairs? So we already know which client it is, we just have to check inside keys of that client which key it is. And I really do not think it is too expensive to check if there is only one key for that user and just returning it, if `kid` is not provided?

**Comment by aeneasr** [comment]
Right, I missed that we can sort/filter by client! In that case your solution makes sense in my opinion!

**Comment by mitar** [comment]
Do we want to make this a strategy so that people can register their own ways to pick up the key? And we can then iterate over them until one returns a key? We could then in fact have two standard key picking strategies? One for RSA and one for ECSDA?

**Comment by aeneasr** [comment]
Yup, that makes sense to me!

**Comment by mitar** [comment]
I think part of this was done in https://github.com/ory/fosite/pull/526/commits/12c4e188ed5a06917bc11f5d97d00da5c2f5c926, but I do not think it is right now. Now if kid is not there it just picks the first signing key from keys. I think this is not the best thing.

Also, I dislike how we have to have hard-coded:

```
		if expectsRSAKey {
			if k, ok := key.Key.(*rsa.PublicKey); ok {
				return k, nil
			}
		} else {
			if k, ok := key.Key.(*ecdsa.PublicKey); ok {
				return k, nil
			}
		}
```

The whole `expectsRSAKey` is ugly. So I suggest we eventually do what this issue is proposing: add a strategy extension for one to implement how a key is picked. You can have then a strict `kid` is required strategy, a strategy which pick RSA key, one which picks ECSDA key, one which picks the first key, and so on.

**Comment by aeneasr** [comment]
A strict `kid` strategy will break OIDC compliance so I would suggest against it. The hard-coding is ugly, that is true, but it is required to prevent JWT alg attacks. The commit also made it to master but with a different hash: https://github.com/ory/fosite/commit/27cc5c0e935ecb8bca23dd8c2670c8a93f7b829d

Closing this as it has been resolved!

**Comment by mitar** [comment]
Yes, but picking the first key from all keys is not the right thing to do. I think this should be further improved, even if not made as a configurable strategy. See what I wrote it the description of this issue:

> But I think key ID in client's public key should not even be required. RFC says that [`kid` header parameter is optional](https://tools.ietf.org/html/rfc7515#section-4.1.4), [key identification section](https://tools.ietf.org/html/rfc7515#section-6) also talks that any way of figuring out the key can be used, and [appendix D](https://tools.ietf.org/html/rfc7515#appendix-D) explicitly mentions simply trying all possibly viable keys.

So if something, the default implementation should try all keys, not pick the first one and if it fails, then abort. I am OK with not making this something we make extendable, but then the default should be that we try all keys and fail only if none of keys succeed (that covers also the case where there are zero keys).

**Comment by mitar** [comment]
The issue why strategies might be useful here is because appendix D is suggesting that key could be picked not just based on `kid`, but also based on `jku`, `jwk`, `x5u`, or `x5c` header parameters. Currently we do not support that if anyone would like to do that (I do not have a need for that, so this is why I think it is OK that we do not have extension at this point in time, until somebody requests it, but I am just noting it down).

**Comment by mitar** [comment]
I do not think fix in 27cc5c0 is a good one.

First, a small nitpick: if `kid` does not exist in the key, then why are error messages at the end of `findPublicKey` function still referencing it?

```
	if expectsRSAKey {
		return nil, errorsx.WithStack(ErrInvalidRequest.WithHintf("Unable to find RSA public key with use='sig' for kid '%s' in JSON Web Key Set.", kid))
	} else {
		return nil, errorsx.WithStack(ErrInvalidRequest.WithHintf("Unable to find ECDSA public key with use='sig' for kid '%s' in JSON Web Key Set.", kid))
	}
```

Second, I think it does not work well. It tries to find the first key in the keyset and returns it, no matter what. What should we be doing instead is to try to validate the token with all keys in turn, until one succeeds. This makes only sense and it enables key rotation. Selecting only the first key effectively makes all other keys useless (except if they contain `kid`), but I think we established that `kid` should not be required.

**Comment by aeneasr** [comment]
> Second, I think it does not work well. It tries to find the first key in the keyset and returns it, no matter what. What should we be doing instead is to try to validate the token with all keys in turn, until one succeeds. This makes only sense and it enables key rotation. Selecting only the first key effectively makes all other keys useless (except if they contain kid), but I think we established that kid should not be required.

Agreed!
